devito/core/gpu_openmp.py

```diff
@@ -349,6 +431,11 @@ def _specialize_iet(cls, graph, **kwargs):
         data_manager.place_definitions(graph)
         data_manager.place_casts(graph)
 
+        # Initialize OpenMP environment
+        initialize(graph)
+        if options['gpu-direct']:
+            mpi_gpu_direct(graph)
```

Comment:
this should be right after `mpiize`

Reply:
Calling `mpi_gpu_direct` before "Symbol definitions" block would create `Blocks` before creating C variables. That would lead to `MPI_Request` variables being local to their blocks. This way, it would generate incorrect C code.

Reply:
I might have a fix in a branch of mine to this issue. Can you leave a TODO comment? Along the lines

```
# TODO: This should be moved right below the `mpiize` pass, but currently ... <same explanation you wrote above>
if options[...
```

Reply:
All right